<a href="https://colab.research.google.com/github/sohanasheikh21/Spam-email-detection/blob/main/spam_email_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

df = pd.read_csv('/content/emails.csv')

print(df.head())

                                                text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1


In [ ]:
print("Dataset shape:", df.shape)
print("\nColumn names:")
print(df.columns)

print("\nSpam and Ham counts:")
print(df['spam'].value_counts())

Dataset shape: (5728, 2)

Column names:
Index(['text', 'spam'], dtype='object')

Spam and Ham counts:
spam
0    4360
1    1368
Name: count, dtype: int64


In [ ]:
print("Missing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

Missing values:
text    0
spam    0
dtype: int64

Duplicate rows:
33


In [ ]:
# Remove duplicate emails
df = df.drop_duplicates()

# Check the new dataset size
print("Dataset shape after removing duplicates:", df.shape)

# Check duplicates again
print("Duplicates remaining:", df.duplicated().sum())

Dataset shape after removing duplicates: (5695, 2)
Duplicates remaining: 0


In [ ]:
X = df['text']
y = df['spam']

print("Email text:")
print(X.head())

print("\nLabels:")
print(y.head())

Email text:
0    Subject: naturally irresistible your corporate...
1    Subject: the stock trading gunslinger  fanny i...
2    Subject: unbelievable new homes made easy  im ...
3    Subject: 4 color printing special  request add...
4    Subject: do not have money , get software cds ...
Name: text, dtype: object

Labels:
0    1
1    1
2    1
3    1
4    1
Name: spam, dtype: int64


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print("Training emails:", len(X_train))
print("Testing emails:", len(X_test))

Training emails: 4556
Testing emails: 1139


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(
    lowercase=True,
    stop_words='english',
    max_features=5000
)

X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)

print("Training data shape:", X_train_tfidf.shape)
print("Testing data shape:", X_test_tfidf.shape)

Training data shape: (4556, 5000)
Testing data shape: (1139, 5000)


In [ ]:
from sklearn.naive_bayes import MultinomialNB

# Create the model
model = MultinomialNB()

# Train the model
model.fit(X_train_tfidf, y_train)

print("Model training completed successfully!")

Model training completed successfully!


In [ ]:
# Predict the test emails
y_pred = model.predict(X_test_tfidf)

print("Predictions completed!")
print(y_pred[:20])

Predictions completed!
[0 0 0 0 0 0 0 0 1 1 0 0 0 0 1 0 1 0 0 0]


In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Predict on test data
y_pred = model.predict(X_test_tfidf)

# Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Confusion Matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# Classification Report
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.9833187006145742

Confusion Matrix:
[[861   4]
 [ 15 259]]

Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       865
           1       0.98      0.95      0.96       274

    accuracy                           0.98      1139
   macro avg       0.98      0.97      0.98      1139
weighted avg       0.98      0.98      0.98      1139



In [ ]:
# Test the model with a new email

new_email = ["Hi, can you please send me the notes for today's lecture?"]

# Convert the new email into TF-IDF format
new_email_tfidf = vectorizer.transform(new_email)

# Predict
prediction = model.predict(new_email_tfidf)

# Display result
if prediction[0] == 1:
    print("Prediction: SPAM")
else:
    print("Prediction: HAM (Not Spam)")

Prediction: HAM (Not Spam)


# SPAM EMAIL DETECTION USING MACHINE LEARNING

**Name:** Sohana Sheikh  
**Roll No:** 13  
**Branch:** AI & DS

### Project Objective
To develop a machine learning model that can classify emails as Spam or Ham (Not Spam).

## 1. Introduction

Spam emails are unwanted messages that may contain advertisements, scams,
fraudulent links or other unwanted content.

In this project, Machine Learning is used to automatically classify emails
into two categories:

• Spam – unwanted email
• Ham – normal or legitimate email

## 2. Dataset

The dataset contains email messages along with their corresponding labels.

• Total emails: 5728
• Spam emails: 1368
• Ham emails: 4360
• Features: Email text
• Target: Spam label

## 3. Data Preprocessing

The dataset was checked for missing values and duplicate records.
Duplicate records were removed before training the model.

## 4. Feature Extraction Using TF-IDF

TF-IDF (Term Frequency-Inverse Document Frequency) converts the email
text into numerical features that can be processed by a machine learning model.

## 5. Model Training

The Multinomial Naive Bayes algorithm was used to train the spam email
classification model.

## 6. Model Evaluation

The trained model was evaluated using accuracy, confusion matrix,
precision, recall and F1-score.

## 7. Testing With New Emails

The trained model was tested with new email messages to check whether
they were classified as Spam or Ham.

## 8. Conclusion

The Spam Email Detection model successfully classified emails into Spam
and Ham categories.

The model achieved an accuracy of 98.33% on the test dataset.

The project demonstrates how Machine Learning and TF-IDF can be used
for automatic email classification.